# SQL Part 3 Studio

### Let's practice your SQL quering skills!  For each question, work along in the notebook adding your query and answering the questions. 

### **DataSet**

### <span style="font-size: 14px; font-family: -apple-system, BlinkMacSystemFont, sans-serif;">We will be using the Goodbooks-10k dataset from the exercises in the prep work.&nbsp; Feel free reference your previous notebooks.</span>

- The dataset can be found here: [goodbooks-10k](https://www.kaggle.com/zygmunt/goodbooks-10k)
- You can access the **BooksDB** in the Launchcode server.

### **Business Issue**

You are an employee at a small independet bookstore and you have been tasked to increase sales.  You decide to use the BooksDB to find books and themes to highlight in fun promotions throughout each month / season.  We will be starting with March and then you will have a chance to explore another month of your choosing.  We want to come up with a list of events/ promotions to run each month.  If you are looking for ideas to include in your promotions:

- [https://www.calendarr.com/united-states/calendar-2022/](https://www.calendarr.com/united-states/calendar-2022/)
- [https://www.holidayinsights.com/moreholidays/](https://www.holidayinsights.com/moreholidays/)
    - Click on a month and it will take you to a more detailed page

## Part 1:  March - Women's History Month, Dr Seuss Day(3/2), National Pie Day3/14, St Patrick's Day (3/17), Season - Spring

### **Event 1:** Women's History Month

Goal: Highlight Popular Womans Authors based on ratings from the Good Reads DB.

### **Write a query that returns tag\_id, the number of times each tag\_id is used and the tag\_name.**  

### Use the GROUP BY and HAVING clause to narrow your focus.

Try multiple keywords, ie woman vs female (female ended up giving better results)

In [9]:
-- Solution
SELECT TOP 10 t.tag_id, COUNT(*) AS 'Number of times tag_id was used', t.tag_name, r.rating
FROM BooksDB.dbo.tags as t
INNER JOIN BooksDB.dbo.ratings as r
ON t.tag_id=r.user_id
GROUP BY t.tag_name, r.rating, t.tag_id
HAVING t.tag_name LIKE '%woman%' OR t.tag_name LIKE '%female%'
ORDER BY COUNT(t.tag_id) DESC, r.rating DESC

(10 rows affected)

Total execution time: 00:00:00.539

tag_id Number of times tag_id was used tag_name rating 11644 55 female-authors 4 28831 54 strong-female-character 4 28834 44 strong-female-lead 5 11644 40 female-authors 5 28831 38 strong-female-character 3 11649 35 female-fiction 5 23891 34 policewoman-potential-victim 5 28831 32 strong-female-character 5 22270 30 older-woman 4 11662 28 female-writers 4

In [8]:
SELECT TOP 10  SUM (bt.count) AS tag_count, t.tag_id, t.tag_name
FROM BooksDB.dbo.tags AS t
LEFT JOIN BooksDB.dbo.book_tags AS bt
ON t.tag_id = bt.tag_id
GROUP BY t.tag_id, t.tag_name
HAVING tag_name LIKE '%woman%' OR tag_name LIKE '%female%' OR tag_name LIKE '%trans' AND tag_name LIKE '%author%'
ORDER BY tag_count DESC;

(10 rows affected)

Total execution time: 00:00:00.414

tag_count tag_id tag_name 17999 11643 female-author 13690 11644 female-authors 2554 11658 female-protagonist 302 11646 female-detective 155 11656 female-pov 147 11662 female-writers 136 32681 woman-s-murder-club 133 11642 female 97 22271 older-woman-younger-man 93 19038 man-woman

### **Which keyword(s) and tag(s) would you like to explore more?**

female-authors : tag\_id = 11644

strong-female-lead: tag\_id = 28834

### **Write a query to return authors, titles, ratings and** **selected tag(s)\_id.**

In [33]:
-- Solution
--A Query for the femle-authors tag
SELECT TOP 10 b.title, b.authors,  bt.tag_id AS 'Female-Authors Tag',bt.[count]
FROM BooksDB.dbo.books as b
INNER JOIN BooksDB.dbo.book_tags as bt
ON b.best_book_id=bt.goodreads_book_id
WHERE bt.tag_id = 11644
ORDER BY bt.[count] DESC

--A Query for the strong-female-lead tag
SELECT TOP 10 b.title, b.authors, bt.tag_id AS 'Strong Female Lead',bt.[count]
FROM BooksDB.dbo.books as b
INNER JOIN BooksDB.dbo.book_tags as bt
ON b.book_id=bt.goodreads_book_id
WHERE bt.tag_id = 32681
ORDER BY bt.[count] DESC


(10 rows affected)

(10 rows affected)

Total execution time: 00:00:00.348

title authors Female-Authors Tag count The Handmaid's Tale Margaret Atwood 11644 214 Jane Eyre Charlotte Brontë, Michael Mason 11644 170 Wuthering Heights Emily Brontë, Richard J. Dunn 11644 158 Frankenstein Mary Wollstonecraft Shelley, Percy Bysshe Shelley, Maurice Hindle 11644 146 The Diary of a Young Girl Anne Frank, Eleanor Roosevelt, B.M. Mooyaart-Doubleday 11644 128 Little Women (Little Women, #1) Louisa May Alcott 11644 120 Beloved Toni Morrison 11644 106 Sense and Sensibility Jane Austen, Tony Tanner, Ros Ballaster 11644 102 The Lovely Bones Alice Sebold 11644 93 Americanah Chimamanda Ngozi Adichie 11644 93

title authors Strong Female Lead count 3rd Degree (Women's Murder Club, #3) James Patterson, Andrew Gross 32681 14 2nd Chance (Women's Murder Club, #2) James Patterson, Andrew Gross 32681 13 7th Heaven (Women's Murder Club, #7) James Patterson, Maxine Paetro 32681 12 The 6th Target (Women's Murder Club, #6) James Patterson, Maxine Paetro 32681 11 The 5th Horseman (Women's Murder Club, #5) James Patterson, Maxine Paetro 32681 11 4th of July (Women's Murder Club, #4) James Patterson, Maxine Paetro 32681 11 The 8th Confession (Women's Murder Club, #8) James Patterson, Maxine Paetro 32681 11 The 9th Judgment (Women's Murder Club, #9) James Patterson, Maxine Paetro 32681 10 11th Hour (Women's Murder Club, #11) James Patterson, Maxine Paetro 32681 9 10th Anniversary (Women's Murder Club, #10) James Patterson, Maxine Paetro 32681 9

In [11]:
SELECT TOP 5 authors, title, average_rating, tag_id 
FROM BooksDB.dbo.books AS b 
LEFT JOIN BooksDB.dbo.book_tags AS bt
ON goodreads_book_id = best_book_id
WHERE tag_id = 32681
ORDER BY average_rating DESC;


(5 rows affected)

Total execution time: 00:00:00.152

authors title average_rating tag_id James Patterson, Maxine Paetro The 9th Judgment (Women's Murder Club, #9) 4.08 32681 James Patterson, Maxine Paetro 7th Heaven (Women's Murder Club, #7) 4.06 32681 James Patterson, Maxine Paetro 10th Anniversary (Women's Murder Club, #10) 4.05 32681 James Patterson, Maxine Paetro 4th of July (Women's Murder Club, #4) 4.04 32681 James Patterson, Maxine Paetro 15th Affair (Women's Murder Club #15) 4.04 32681

### **Record your thoughts about why you wrote the query the way you did.**

# Part 2: Choose Another Month

Choose another month and plan at least 2 events / promotions and answer the following questions.

Which month did you choose?  **Frebruary**

What 2 events / promotions are you highlighting?

> 1. Black History month

> 2. Valentines Day

# Part 3: Summarize your Work

For each event write at least one query that joins any two tables in the BooksDB to support your choice and record you thoughts as to why you used the paticlular query. At least one of your queries needs to include a HAVING clause.

In [34]:
-- BLACK HISTORY MONTH
--Find tags with 'black' or 'african american'
--Im going to use african-american-history (tag_id 1759)
SELECT TOP 10 t.tag_id, COUNT(t.tag_id) AS 'Number of times tag_id was used', t.tag_name, r.rating
FROM BooksDB.dbo.tags as t
INNER JOIN BooksDB.dbo.ratings as r
ON t.tag_id=r.user_id
GROUP BY t.tag_name, r.rating, t.tag_id
HAVING t.tag_name LIKE '%black%' OR t.tag_name LIKE '%african%'
ORDER BY COUNT(t.tag_id) DESC, r.rating DESC

--A Query for the African American History tag
SELECT TOP 10 b.title, b.authors, bt.tag_id AS 'African American History', bt.[count]
FROM BooksDB.dbo.books as b
INNER JOIN BooksDB.dbo.book_tags as bt
ON b.book_id=bt.goodreads_book_id
WHERE bt.tag_id = 1759
ORDER BY bt.[count] DESC

(10 rows affected)

(10 rows affected)

Total execution time: 00:00:00.704

tag_id Number of times tag_id was used tag_name rating 27499 116 shayla-black 5 29703 92 the-chronicle-of-the-black-company 5 4706 52 black-fire 5 4715 50 black-liberation 4 29703 41 the-chronicle-of-the-black-company 3 4715 41 black-liberation 3 4706 40 black-fire 4 1759 40 african-american-history 4 29661 39 the-blackstone-affair 5 4698 39 black-dagger-legacy 3

title authors African American History count The Warmth of Other Suns: The Epic Story of America's Great Migration Isabel Wilkerson 1759 56 Twelve Years a Slave Solomon Northup 1759 32 Narrative of the Life of Frederick Douglass Frederick Douglass 1759 26 The Autobiography of Malcolm X Malcolm X, Alex Haley 1759 24 The Souls of Black Folk W.E.B. Du Bois, Donald B. Gibson, Monica W. Elbert 1759 19 March: Book One (March, #1) John Lewis, Andrew Aydin, Nate Powell 1759 17 Up from Slavery Booker T. Washington, Ishmael Reed 1759 16 Malcolm X: A Life of Reinvention Manning Marable 1759 13 Incidents in the Life of a Slave Girl Harriet Jacobs 1759 12 Cane River Lalita Tademy 1759 11

In [35]:
-- VALENTINES DAY
--Find tags with 'love' or 'romance'
--Im going to use Lesbian Romance (tag_id 17937)
SELECT TOP 10 t.tag_id, COUNT(t.tag_id) AS 'Number of times tag_id was used', t.tag_name, r.rating
FROM BooksDB.dbo.tags as t
INNER JOIN BooksDB.dbo.ratings as r
ON t.tag_id=r.user_id
GROUP BY t.tag_name, r.rating, t.tag_id
HAVING t.tag_name LIKE '%love%' OR t.tag_name LIKE '%romance%'
ORDER BY COUNT(t.tag_id) DESC, r.rating DESC


--A Query for the Lesbian Romance tag
SELECT b.title, b.authors, bt.tag_id AS 'Lesbian Romance Tag', bt.[count]
FROM BooksDB.dbo.books as b
INNER JOIN BooksDB.dbo.book_tags as bt
ON b.book_id=bt.goodreads_book_id
WHERE bt.tag_id = 17937
ORDER BY bt.[count] DESC

(10 rows affected)

(4 rows affected)

Total execution time: 00:00:00.648

tag_id Number of times tag_id was used tag_name rating 26145 141 romance-contemporaneo 3 17937 120 lesbian-romance 5 27612 102 shifter-romance 5 7120 85 christian-fiction-romance 5 26146 82 romance-contemporary 5 17130 78 kids-books-i-love 5 21058 68 na-romance 5 7674 60 college-love 4 20634 57 multiple-lovers 5 31867 55 victorian-romance 4

title authors Lesbian Romance Tag count Tipping the Velvet Sarah Waters 17937 16 Annie on My Mind Nancy Garden 17937 12 Empress of the World (Battle Hall Davies, #1) Sara Ryan 17937 8 Keeping You a Secret Julie Anne Peters 17937 7

### **Summary of event 1:**

Double Click to edit.

- Black History Month: In my search for books that contain the tag **black-american-history** I saw that this tag was used a total of 40 times and had an average rating of 4.
- The Warmth of Other Suns: The Epic Story of America's Great Migration had the most **count of 56**

### **Summary of event 2:**

- Valentines Day: I queried  the tags that has the word 'romance' or 'love' and **lesbian-romance** was the most taged with the highest rating. 
    
- The book with the highest count of that category was **Tipping the Velvet with a count of 16**